perl-lib/OESS/lib/OESS/Database.pm

```diff
@@ -10197,6 +10219,72 @@ sub revert_edge_interface_move_maintenance {
     };
 }
 
+=head2 move_vrf_endpoints
+
+=cut
+
+sub move_vrf_endpoints {
+    my ($self, %args) = @_;
+    my $orig_interface_id  = $args{'orig_interface_id'};
+    my $new_interface_id   = $args{'new_interface_id'};
+    my %used_vlans;
+    
+    # Gather occupied vlans from circuits on the new interface
+    my $dst_edge_interface_recs = $self->get_circuit_edge_interface_memberships(
+        interface_id => $new_interface_id);
+    foreach my $edge_int_rec (@$dst_edge_interface_recs) {
+        $used_vlans{$edge_int_rec->{'external_vlan_id'}} = 1;
+    }
+
+    # Gather occupied vlans from vrf endpoints on the new interface
+    my $query = "SELECT tag FROM vrf_ep ".
+                "WHERE interface_id = ? AND state = 'active'";
+    my $rows = $self->_execute_query($query, [$new_interface_id]);
+    if(!defined($rows)) {
+        $self->_set_error("Error fetching new interface's vrf endpoints");
+        return;
+    }
+    foreach my $row (@$rows){
+        $used_vlans{$rows->{'tag'}} = 1;
+    }
+
+    # Gather the vrf endpoints we want to attempt to move
+    $query = "SELECT vrf_ep_id, tag, inner_tag, unit FROM vrf_ep ".
+             "WHERE interface_id = ? AND state = 'active'";
+    $rows = $self->_execute_query($query, [$orig_interface_id]);
+    if(!defined($rows)) {
+        $self->_set_error("Error fetching old interface's vrf endpoints");
+        return;
+    }
+    foreach my $row (@$rows) {
+        if($used_vlans{$row->{'tag'}}) {
+            next;
+        }
+        
+        # If QnQ, make sure no unit conflicts, or alternatively just generate a new one
+        my $unit_number = $row->{'unit'};
+        if($row->{'inner_tag'} != undef) {
+            $unit_number = $self->find_available_unit(
+                    interface_id => $new_interface_id,
+                    tag          => $row->{'tag'},
+                    inner_tag    => $row->{'inner_tag'});
+        }
+
+        # Update the interface_id (and unit number if needed)
+        $query = "UPDATE vrf_ep SET interface_id = ?, unit = ? ".
```

Comment:
if we use the already existing objects I think we can add this checking in (basically I would rather have our objects do it, so we can use unit tests to validate correctness)

Reply:
By object you mean unit? Are you saying to not use a new unit number if we don't need to?